README.md

```diff
@@ -18,4 +20,214 @@ Execute the following steps to practice various topics covered in the intro to g
+
+![Rebase->Merge State](/content/images/rebase->merge_state.png)
+
+## Cherry-pick
+
+`git cherry-pick` is used to apply the changes introduced by some existing commits. It is useful when you want to apply a single commit from one branch to another.
+
+At this point, if you are on `main`, you will **not** have the `new2.txt` and `new3.txt` files.
+
+If you switch to `mySecondBranch`, you will have the `new2.txt` and `new3.txt` files as well as the `new4.txt` file.
+
+Now, consider the following scenario:
+
+You want the `new3.txt` file to be added to the `main` branch, but you do not want the `new2.txt` file to be added.
+
+To achieve this, you can use the `git cherry-pick` command.
+
+1) Use `git log` to find the commit hash of the commit where you added `new3.txt`.
+
+    In the example below, the commit hash is `c8789dd`.
+
+    ![Commit Log 2](/content/images/log_rebase_2.png)
+
+    Note: The commit hash will be different in your case.
+
+2) Make sure you are on the `main` branch and use the following command to cherry-pick the commit:
+
+    ```bash
+    git cherry-pick <<commit_hash>>
+    ```
+
+3) Verify that the changes have been applied to the `main` branch by using the `git log` command. You can also verify this by checking that you now have the `new3.txt` file in the `main` branch, but not the `new2.txt` file.
+
+    ![Commit Log 3](/content/images/log_cherrypick_1.png)
+
+## Interactive Rebase
+
+`git rebase -i` is used to interactively rebase commits. This is useful when you want to modify, delete, or combine commits.
+
+Consider the following scenario:
+
+1) Switch to `mySecondBranch`. Add a new text file called `new5.txt` and add some text to it. Add and commit these changes.
+2) Edit the `new2.txt` file and add some more text to it. Add and commit these changes.
+3) Now, you should have the following commit history:
+
+    ![Commit Log 4](/content/images/log_irebase.png)
+
+4) You want to rebase `mySecondBranch` onto `main`, but you have already cherry-picked the commit that added `new3.txt` to the `main` branch.
+
+    Now, you want the `new2.txt` file in your main branch, but it has two commits associated with it. You want to combine these two commits into a single commit. Moreover, you do not want the `new5.txt` file anymore.
+
+    Checkout to `mySecondBranch`:
+
+    At this point, if you wanted to modify the commit history of `mySecondBranch`, you can use: `git rebase -i HEAD~4`. Notice that `HEAD` refers to the current commit (the latest commit on `mySecondBranch`), and `HEAD~4` refers to the commit which is four commits **before** the current commit (in this case, the first commit on `mySecondBranch`). This command will not rebase onto `main`, but rather onto the commit denoted by `HEAD~4`. In this case, it will rebase onto itself, effectively allowing you to modify the commit history of `mySecondBranch` by interacting with the 4 commits on `mySecondBranch`.
```

Comment:
```suggestion
    At this point, if you wanted to modify the commit history of `mySecondBranch`, you can use: `git rebase -i HEAD~4`. Notice that `HEAD` refers to the current commit (the latest commit on `mySecondBranch`), and `HEAD~4` refers to the commit which is four commits **before** the current commit (in this case, the first commit on `mySecondBranch`). This command will not rebase onto `main`, but rather onto the commit denoted by `HEAD~4`. In this case, it will rebase onto itself, effectively allowing you to modify the commit history of `mySecondBranch` by interacting with the last 4 commits on `mySecondBranch`. 
Note that you need to be careful to not rebase with a number greater than 4 if you have committed for a maximum of 4 times in the current PR, else the commits in `main` before this PR will also be pulled in and will clutter the repository
```